test/core/startup.test.js

```diff
@@ -0,0 +1,21 @@
+/* global globalThis */
+
+"use strict";
+
+describe("startup", function() {
+  it("will work when Set is not in the environment", function() {
+    if (typeof require !== "undefined") {
+      var origSet = globalThis.Set;
+      // delete parsimmon from require cache and delete Set from environment
+      delete require.cache[require.resolve("../..")];
+      delete globalThis.Set;
+      // require again, triggers Parsimmon without Set
+      globalThis.Parsimmon = require("../..");
+      assert.isNotOk(globalThis.Parsimmon.Set);
+      // go back to original state
+      delete require.cache[require.resolve("../..")];
```

Comment:
Would it be possible to do `Parsimmon.Set = undefined` and `Parsimmon.Set = origSet` rather than busting the require cache and reloading it? Seems like it would be shorter, but I'm not sure if there's a caveat I'm missing.

Reply:
This is basically testing https://github.com/jneen/parsimmon/pull/298/files#diff-d6843052e5b691e32d17eb1fb2508987R1402

This is testing the initial require of `Parsimmon` when `Set` is `undefined`. That last `: undefined` is only run during require.

Reply:
Oh, that makes sense. I remember this was a bit of a sticking point with the Binary module inside Parsimmon. We solved that with a helper function to check instead of hard coding it at initialization (as Chai apparently does as well).

```js
function bufferExists() {
  return typeof Buffer !== "undefined";
}
```

That way we can test the code path by just stubbing it temporarily:

```js
  context("Buffer is not present.", function() {
    var buff;
    before(function() {
      buff = global.Buffer;
      global.Buffer = undefined;
    });

    after(function() {
      global.Buffer = buff;
    });

    it("Disallows construction.", function() {
      assert.throws(function() {
        Parsimmon.Binary.bitSeqObj(0xf);
      }, /buffer global/i);
    });
  });
```

Reply:
I don't think that will work with chai. That was my original solution (because the check for `Set` was in the function it was essentially lazy) .. but the problem was that `chai` itself reads `Set` when it is required, and then when you assert with `deepEqual` against an object, it throws.

```
TypeError: Cannot read property 'prototype' of undefined
      at typeDetect (node_modules/type-detect/type-detect.js:285:41)
      at extensiveDeepEqual (node_modules/deep-eql/index.js:183:22)
      at Object.deepEqual [as eql] (node_modules/deep-eql/index.js:106:10)
      at Proxy.assertEql (node_modules/chai/lib/chai/core/assertions.js:1085:11)
      at Proxy.methodWrapper (node_modules/chai/lib/chai/utils/addMethod.js:57:25)
      at Function.assert.deepEqual.assert.deepStrictEqual (node_modules/chai/lib/chai/interface/assert.js:232:56)
      at Context.<anonymous> (test/core/seq.test.js:73:12)
      at processImmediate (internal/timers.js:458:21)
```

Alternately, I could essentially leave the runtime code as it was (check for `Set` inside the function using it), and then use `require.cache` and `require` to bust *chai*s cache instead and force it to load thinking `Set` is not in the environment, that should allow for something like above.

Thoughts?

Reply:
Sorry I didn't quite explain that fully, but I meant that approach *plus* using `Parsimmon._Set` to store the value so we can dependency inject it inside tests, thus not mucking up Chai.

Reply:
Ok, I think I see where you are going with this. Basically lazily set whether `Set` exists to avoid having to test the initial setting that is happening now during `require`. We don't actually need to store a reference to `Set`, just whether we found it exists.

```javascript
function setExists() {
  if (Parsimmon._supportsSet !== "undefined") {
    return Parsimmon._supportsSet;
  }
  var exists = typeof Set !== "undefined";
  Parsimmon._supportsSet = exists;
  return exists;
}
```

Both branches of code will be tested as we test these scenarios. Sound good?

Reply:
Yeah, I think we're on the same page now. Thanks for your patience with testing on this!